File: studio-frontend-prototype/src/compose.ts
//! Match what a product needs against the components this system knows.
//!
//! Shared by the App Spec's Compose button and the project's Components tab,
//! which ask the same question from two directions and must not answer it two
//! ways.
//!
//! ── Why a candidate carries whether it was ever built ─────────────────────────
//!
//! A gear directory in the catalogue is not a gear. Interviewing Acronis
//! (2026-09-18) the point was made while reading the repository listing on
//! screen — "тут есть документы, дизайн, но ничего нету, реализации никакой
//! нет. Вот как это считать?" — against a roadmap number roughly five times the
//! count of gears anyone has finished.
//!
//! That is not a complaint about the catalogue, it is a defect in any tool that
//! reads it. Ranking by keyword alone puts a well-written stub above a shipped
//! component, because a stub is mostly prose and prose is what the keywords
//! match. Suggesting it is worse than suggesting nothing: it answers "what can
//! we build this from?" with something nobody can build from.
//!
//! The catalogue already knows. The repository scan counts the crates under a
//! gear's directory (`components_catalog/repo_enrich.rs`), and on this stand ten
//! of forty-two Rust gears count zero — `approval-service` among them, which is
//! one of the two he named. So candidates are sorted built-first and the rest
//! are labelled, rather than quietly dropped: a design may legitimately name a
//! component that is still only a design.

import type { CatalogNode, Capability } from "./api";

/** What the catalogue can say about whether a component was ever built.
 *
 *  `null` is not a maybe — it means the question does not apply or was never
 *  asked. FrontX packages carry no crate count at all, and a component with no
 *  profile has not been scanned. Neither is evidence of absence, so neither is
 *  reported as one. */
export type BuildState = "built" | "docs-only" | null;

export type Candidate = {
  name: string;
  kind: string;
  score: number;
  why: string[];
  built: BuildState;
};

export type PlanRow = {
  capability: string;
  candidates: Candidate[];
  /** No candidate at all — the capability has nothing to build from. */
  gap: boolean;
  /** Candidates exist, but none of them has been built. Not a gap, and not an
   *  answer either: worth saying out loud rather than leaving to the reader. */
  unbuilt: boolean;
};

function profileText(profile?: Record<string, unknown>): string {
  const auto = profile?.auto;
  if (auto && typeof auto === "object") {
    const d = (auto as Record<string, unknown>).description;
    if (d && typeof d === "object") {
      const s = (d as Record<string, unknown>).s;
      if (typeof s === "string") return s;
    }
    if (typeof d === "string") return d;
  }
  return "";
}

/** Crates under the component's directory, as the repository scan counted them.
 *
 *  Read from `auto.crates.n`, which the scan writes for a Rust gear and omits
 *  for anything it did not scan that way. Zero is the load-bearing value: a
 *  directory holding `docs/` and `gear.toml` and no crate. */
export function buildState(profile?: Record<string, unknown>): BuildState {
  const auto = profile?.auto;
  if (!auto || typeof auto !== "object") return null;
  const crates = (auto as Record<string, unknown>).crates;
  if (!crates || typeof crates !== "object") return null;
  const n = (crates as Record<string, unknown>).n;
  if (typeof n !== "number") return null;
  return n > 0 ? "built" : "docs-only";
}

/** Does the text mention `term` where a word begins?
 *
 *  `includes` was the old test, and on the live catalogue it offered
 *  `@gears-frontx/framework` for the `rating` capability — because its
 *  description says "FrontX framework integ**rating** all SDK packages". A
 *  suggestion list is a claim that these components fit, and a coincidence of
 *  spelling is the cheapest possible way to break that claim.
 *
 *  Only the START of the term is anchored, which is not the obvious rule and is
 *  the one the catalogue asked for. Every false positive measured there was the
 *  term swallowed by the END of a longer word — `source` in "resource", `file`
 *  in "profile", `entity` in "machine-identity", `graph` in "cryptography".
 *  Every genuine match lost to a both-ends anchor was the term BEGINNING one —
 *  `deploy` in "deployment-topology", `node` in "nodes-registry",
 *  `subscription` in "subscriptions". Anchoring the head alone keeps the second
 *  set and drops the first.
 *
 *  The boundary is "not a letter or digit" rather than `\b`, so a hyphen, a
 *  slash and an `@` all start a word: `storage` finds `cf-gears-file-storage`
 *  and `state` finds `@gears-frontx/state`. */
function mentions(hay: string, term: string): boolean {
  const t = term.trim().toLowerCase();
  if (!t) return false;
  const escaped = t.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`(?<![a-z0-9])${escaped}`).test(hay);
}

function componentHaystack(g: CatalogNode, profile?: Record<string, unknown>): string {
  const v = g.value;
  return [
    v.name ?? "",
    v.description ?? "",
    v.kind ?? "",
    (v.keywords ?? []).join(" "),
    (v.categories ?? []).join(" "),
    profileText(profile),
  ]
    .filter(Boolean)
    .join(" ")
    .toLowerCase();
}

/** Resolve capabilities to candidate components.
 *
 *  `vocabulary` is the workspace's effective capability catalogue, which used to
 *  be a `CAP_KEYWORDS` constant in this file. A workspace that invents a
 *  capability can now give it search terms instead of getting zero candidates
 *  and no explanation (ADR-0014 s5). A capability the catalogue does not know is
 *  still matched against its own name, exactly as before.
 *
 *  Candidates come back built-first, then by score. The cut to five happens
 *  after that sort, so a shipped component is never displaced from the list by
 *  a stub that mentioned the word more often. */
export function composePlan(
  caps: string[],
  gears: CatalogNode[],
  profiles: Record<string, Record<string, unknown>>,
  vocabulary: readonly Capability[],
): PlanRow[] {
  const terms = new Map(vocabulary.map((c) => [c.key, c.terms]));
  const geared = gears.filter((g) => typeof g.value.name === "string");
  const rank = (c: Candidate) => (c.built === "built" ? 0 : c.built === null ? 1 : 2);
  return caps.map((cap) => {
    const kws = terms.get(cap)?.length ? terms.get(cap)! : [cap];
    const candidates = geared
      .map((g) => {
        const profile = profiles[g.value.name as string];
        const hay = componentHaystack(g, profile);
        const why = new Set<string>();
        for (const k of kws) if (mentions(hay, k)) why.add(k);
        if (mentions(hay, cap)) why.add(cap);
        return {
          name: g.value.name as string,
          kind: g.value.kind ?? "gear",
          score: why.size,
          why: Array.from(why),
          built: buildState(profile),
        };
      })
      .filter((c) => c.score > 0)
      .sort((a, b) => rank(a) - rank(b) || b.score - a.score)
      // One component, one candidate. Sixteen of the 118 names on this stand are
      // stored twice -- the same FrontX package under both
      // `catalog.frontx.v1` and `catalog.gear.v1` -- and a list of five that
      // spends two slots on one package is offering four. Deduplication comes
      // after the sort, so the copy that survives is the better-ranked one.
      .filter((c, i, all) => all.findIndex((o) => o.name === c.name) === i)
      .slice(0, 5);
    return {
      capability: cap,
      candidates,
      gap: candidates.length === 0,
      unbuilt: candidates.length > 0 && candidates.every((c) => c.built === "docs-only"),
    };
  });
}

/** The profiles keyed the way `composePlan` wants them: by component name.
 *
 *  Both callers were doing this inline, and one of them reading `gear_name` off
 *  a node whose other half is a `uml` array is not obvious enough to write
 *  twice. */
export function profilesByName(nodes: CatalogNode[]): Record<string, Record<string, unknown>> {
  const out: Record<string, Record<string, unknown>> = {};
  for (const n of nodes) {
    const name = (n.value as Record<string, unknown>).gear_name;
    if (typeof name === "string") out[name] = n.value as Record<string, unknown>;
  }
  return out;
}

File: studio-backend/src/components_catalog/compose.rs
//! Matching what a product needs against the components this system knows.
//!
//! Asked from two directions — the App Spec's Compose button and a project's
//! Components tab — and it must not answer them two ways. That was already the
//! reason this was one module in the portal rather than two call sites; with a
//! second portal arriving it stops being enough, because the two would make
//! their own copies of the rules below and disagree quietly.
//!
//! ── Why a candidate carries whether it was ever built ────────────────────────
//!
//! A gear directory in the catalogue is not a gear. Reading the repository
//! listing on screen during an interview (2026-09-18) the point was made
//! plainly — there are documents and design in there and no implementation at
//! all, so how should that count? — against a roadmap number roughly five times
//! the count of gears anyone has finished.
//!
//! That is not a complaint about the catalogue. It is a defect in any tool that
//! reads it and ranks by keyword alone, because a well-written stub is mostly
//! prose and prose is what keywords match. Suggesting it is worse than
//! suggesting nothing: it answers "what can we build this from?" with something
//! nobody can build from.
//!
//! So candidates are sorted built-first and the rest are LABELLED rather than
//! dropped — a design may legitimately name a component that is still only a
//! design — and the cut to a handful happens after that sort, so a shipped
//! component is never displaced from the list by a stub that mentioned the word
//! more often.

use serde_json::Value;

/// What the catalogue can say about whether a component was ever built.
///
/// `Unknown` is not a maybe. It means the question does not apply or was never
/// asked: a FrontX package carries no crate count at all, and a component with
/// no profile has not been scanned. Neither is evidence of absence, so neither
/// is reported as one.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BuildState {
    Built,
    DocsOnly,
    Unknown,
}

impl BuildState {
    pub fn as_str(self) -> &'static str {
        match self {
            BuildState::Built => "built",
            BuildState::DocsOnly => "docs-only",
            BuildState::Unknown => "unknown",
        }
    }

    /// Built first, then the unscanned, then what is known to be docs only.
    ///
    /// The unknown sits in the middle deliberately: it might be built, and
    /// ranking it below something known NOT to be would be asserting more than
    /// the catalogue said.
    fn rank(self) -> u8 {
        match self {
            BuildState::Built => 0,
            BuildState::Unknown => 1,
            BuildState::DocsOnly => 2,
        }
    }
}

/// One component offered for one capability.
#[derive(Debug, Clone, PartialEq)]
pub struct Candidate {
    pub name: String,
    pub kind: String,
    /// How many of the capability's terms this component mentions.
    pub score: usize,
    /// Which terms they were — the reason, so a suggestion can be argued with.
    pub why: Vec<String>,
    pub built: BuildState,
}

/// One capability, and what could fill it.
#[derive(Debug, Clone, PartialEq)]
pub struct PlanRow {
    pub capability: String,
    pub candidates: Vec<Candidate>,
    /// No candidate at all — the capability has nothing to build from.
    pub gap: bool,
    /// Candidates exist, but none has been built. Not a gap, and not an answer
    /// either: worth saying out loud rather than leaving to the reader.
    pub unbuilt: bool,
}

/// How many candidates a row offers before the tail is cut.
const SHORTLIST: usize = 5;

/// Does `hay` mention `term` at the start of a word?
///
/// ANCHORED AT THE HEAD ONLY, and that asymmetry is the whole rule. Anchoring
/// both ends loses the matches that are genuine — `deploy` in
/// `deployment-topology`, `node` in `nodes-registry`, `subscription` in
/// `subscriptions`. Anchoring neither end gains the matches that are noise —
/// `source` swallowed by `resource`, `file` by `profile`, `entity` by
/// `machine-identity`, `graph` by `cryptography`. Every false positive found in
/// practice was a term ending inside a longer word; every genuine match lost to
/// a both-ends anchor was a term beginning one.
///
/// The boundary is "not a letter or digit" rather than a word boundary, so a
/// hyphen, a slash and an `@` all start a word: `storage` finds
/// `cf-gears-file-storage` and `state` finds `@gears-frontx/state`.
fn mentions(hay: &str, term: &str) -> bool {
    let term = term.trim().to_lowercase();
    if term.is_empty() {
        return false;
    }
    let hay_bytes = hay.as_bytes();
    let mut from = 0usize;
    while let Some(found) = hay[from..].find(&term) {
        let at = from + found;
        let preceded_by_word_char = at > 0
            && hay_bytes
                .get(at - 1)
                .is_some_and(|b| b.is_ascii_alphanumeric());
        if !preceded_by_word_char {
            return true;
        }
        from = at + 1;
        if from >= hay.len() {
            break;
        }
    }
    false
}

/// Everything about a component that a capability's terms are matched against.
fn haystack(component: &Value, profile: Option<&Value>) -> String {
    let text = |key: &str| {
        component
            .get(key)
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned()
    };
    let list = |key: &str| {
        component
            .get(key)
            .and_then(Value::as_array)
            .map(|items| {
                items
                    .iter()
                    .filter_map(Value::as_str)
                    .collect::<Vec<_>>()
                    .join(" ")
            })
            .unwrap_or_default()
    };
    [
        text("name"),
        text("description"),
        text("kind"),
        list("keywords"),
        list("categories"),
        profile_text(profile),
    ]
    .join(" ")
    .to_lowercase()
}

/// The description the repository scan wrote, wherever it put it.
fn profile_text(profile: Option<&Value>) -> String {
    let Some(auto) = profile.and_then(|p| p.get("auto")) else {
        return String::new();
    };
    let Some(description) = auto.get("description") else {
        return String::new();
    };
    // The scan has written both a bare string and `{ s: "…" }`.
    description
        .get("s")
        .and_then(Value::as_str)
        .or_else(|| description.as_str())
        .unwrap_or_default()
        .to_owned()
}

/// Resolve capabilities to candidate components.
///
/// `terms` is the workspace's effective capability vocabulary — a capability it
/// invented can give its own search terms rather than getting zero candidates
/// and no explanation (ADR-0014 §5). A capability the vocabulary does not know
/// is matched against its own name, which is what it meant before vocabularies
/// existed.
pub fn plan(
    capabilities: &[String],
    components: &[Value],
    profiles: &serde_json::Map<String, Value>,
    terms: &std::collections::BTreeMap<String, Vec<String>>,
) -> Vec<PlanRow> {
    capabilities
        .iter()
        .map(|capability| {
            let own = vec![capability.clone()];
            let words = terms
                .get(capability)
                .filter(|t| !t.is_empty())
                .unwrap_or(&own);
            let mut candidates: Vec<Candidate> = components
                .iter()
                .filter_map(|component| {
                    let name = component.get("name").and_then(Value::as_str)?;
                    let hay = haystack(component, profiles.get(name));
                    let mut why: Vec<String> = Vec::new();
                    for word in words.iter().chain(std::iter::once(capability)) {
                        if mentions(&hay, word) && !why.iter().any(|w| w == word) {
                            why.push(word.clone());
                        }
                    }
                    if why.is_empty() {
                        return None;
                    }
                    Some(Candidate {
                        name: name.to_owned(),
                        kind: component
                            .get("kind")
                            .and_then(Value::as_str)
                            .unwrap_or("gear")
                            .to_owned(),
                        score: why.len(),
                        why,
                        built: build_state(profiles.get(name)),
                    })
                })
                .collect();
            // Built first, then by score. The cut comes AFTER, so a shipped
            // component is never displaced by a stub that said the word more.
            candidates.sort_by(|a, b| {
                a.built
                    .rank()
                    .cmp(&b.built.rank())
                    .then(b.score.cmp(&a.score))
                    .then(a.name.cmp(&b.name))
            });
            let unbuilt =
                !candidates.is_empty() && !candidates.iter().any(|c| c.built == BuildState::Built);
            candidates.truncate(SHORTLIST);
            PlanRow {
                capability: capability.clone(),
                gap: candidates.is_empty(),
                unbuilt,
                candidates,
            }
        })
        .collect()
}

/// What the catalogue says about a component, when it says anything.
///
/// The scan computes `gear_status` now, so every consumer gets one answer
/// instead of deriving its own. Reading the crate count stays as the fallback:
/// a graph synced before the status existed carries the count and not the word.
fn build_state(profile: Option<&Value>) -> BuildState {
    let Some(profile) = profile else {
        return BuildState::Unknown;
    };
    if let Some(status) = profile
        .get("auto")
        .and_then(|a| a.get("gear_status"))
        .and_then(Value::as_str)
    {
        return match status {
            "built" => BuildState::Built,
            "docs-only" => BuildState::DocsOnly,
            _ => BuildState::Unknown,
        };
    }
    match profile
        .get("auto")
        .and_then(|a| a.get("crates"))
        .and_then(|c| c.get("n"))
        .and_then(Value::as_u64)
    {
        // Zero is the load-bearing value: a directory holding `docs/` and
        // `gear.toml` and no crate at all.
        Some(0) => BuildState::DocsOnly,
        Some(_) => BuildState::Built,
        None => BuildState::Unknown,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn component(name: &str, description: &str) -> Value {
        json!({ "name": name, "kind": "gear", "description": description })
    }

    fn vocabulary(pairs: &[(&str, &[&str])]) -> std::collections::BTreeMap<String, Vec<String>> {
        pairs
            .iter()
            .map(|(k, terms)| {
                (
                    (*k).to_owned(),
                    terms.iter().map(|t| (*t).to_owned()).collect(),
                )
            })
            .collect()
    }

    // ---- the matching rule ----------------------------------------------

    #[test]
    fn a_term_beginning_a_longer_word_matches() {
        // The set a both-ends anchor would have lost.
        assert!(mentions("deployment-topology", "deploy"));
        assert!(mentions("nodes-registry", "node"));
        assert!(mentions("subscriptions", "subscription"));
    }

    #[test]
    fn a_term_swallowed_by_a_longer_word_does_not() {
        // The set no anchor at all would have gained.
        assert!(!mentions("resource", "source"));
        assert!(!mentions("profile", "file"));
        assert!(!mentions("machine-identity", "entity"));
        assert!(!mentions("cryptography", "graph"));
    }

    #[test]
    fn punctuation_starts_a_word() {
        assert!(mentions("cf-gears-file-storage", "storage"));
        assert!(mentions("@gears-frontx/state", "state"));
    }

    #[test]
    fn an_empty_term_matches_nothing() {
        assert!(!mentions("anything at all", "   "));
    }

    // ---- ranking ---------------------------------------------------------

    /// The rule the interview produced: a stub does not outrank a shipped
    /// component, however many times it says the word.
    #[test]
    fn a_built_component_outranks_a_stub_with_a_better_score() {
        let components = vec![
            component("chatty-stub", "chat chat chat messaging"),
            component("real-chat", "chat"),
        ];
        let profiles: serde_json::Map<String, Value> = [
            (
                "chatty-stub".to_owned(),
                json!({ "auto": { "crates": { "n": 0 } } }),
            ),
            (
                "real-chat".to_owned(),
                json!({ "auto": { "crates": { "n": 3 } } }),
            ),
        ]
        .into_iter()
        .collect();
        let rows = plan(
            &["chat".to_owned()],
            &components,
            &profiles,
            &vocabulary(&[("chat", &["chat", "messaging"])]),
        );
        assert_eq!(rows[0].candidates[0].name, "real-chat");
        assert_eq!(rows[0].candidates[0].built, BuildState::Built);
        // The stub is kept and labelled, not dropped: a design may name a
        // component that is still only a design.
        assert_eq!(rows[0].candidates[1].built, BuildState::DocsOnly);
        assert!(!rows[0].unbuilt);
    }

    #[test]
    fn a_capability_nothing_matches_is_a_gap() {
        let rows = plan(
            &["telepathy".to_owned()],
            &[component("chat", "messaging")],
            &serde_json::Map::new(),
            &vocabulary(&[]),
        );
        assert!(rows[0].gap);
        assert!(!rows[0].unbuilt);
        assert!(rows[0].candidates.is_empty());
    }

    /// Not a gap, and not an answer either.
    #[test]
    fn candidates_that_are_all_unbuilt_are_said_out_loud() {
        let profiles: serde_json::Map<String, Value> = [(
            "stub".to_owned(),
            json!({ "auto": { "crates": { "n": 0 } } }),
        )]
        .into_iter()
        .collect();
        let rows = plan(
            &["chat".to_owned()],
            &[component("stub", "chat")],
            &profiles,
            &vocabulary(&[]),
        );
        assert!(!rows[0].gap);
        assert!(rows[0].unbuilt);
    }

    #[test]
    fn a_capability_the_vocabulary_does_not_know_is_matched_on_its_own_name() {
        let rows = plan(
            &["billing".to_owned()],
            &[component("billing-engine", "invoices")],
            &serde_json::Map::new(),
            &vocabulary(&[]),
        );
        assert_eq!(rows[0].candidates[0].name, "billing-engine");
        assert_eq!(rows[0].candidates[0].why, vec!["billing".to_owned()]);
    }

    #[test]
    fn the_shortlist_is_cut_after_the_sort_not_before() {
        // Six stubs that all match, and one built component last in input
        // order. A cut before the sort would drop the only usable answer.
        let mut components: Vec<Value> = (0..6)
            .map(|i| component(&format!("stub-{i}"), "chat chat"))
            .collect();
        components.push(component("shipped", "chat"));
        let mut profiles = serde_json::Map::new();
        for i in 0..6 {
            profiles.insert(
                format!("stub-{i}"),
                json!({ "auto": { "crates": { "n": 0 } } }),
            );
        }
        profiles.insert(
            "shipped".to_owned(),
            json!({ "auto": { "crates": { "n": 2 } } }),
        );
        let rows = plan(
            &["chat".to_owned()],
            &components,
            &profiles,
            &vocabulary(&[]),
        );
        assert_eq!(rows[0].candidates.len(), SHORTLIST);
        assert_eq!(rows[0].candidates[0].name, "shipped");
    }

    // ---- build state -----------------------------------------------------

    #[test]
    fn the_scans_own_word_is_preferred_to_the_crate_count() {
        let profile = json!({ "auto": { "gear_status": "docs-only", "crates": { "n": 9 } } });
        assert_eq!(build_state(Some(&profile)), BuildState::DocsOnly);
    }

    #[test]
    fn no_profile_is_unknown_rather_than_docs_only() {
        // Never scanned is not the same fact as scanned and found empty.
        assert_eq!(build_state(None), BuildState::Unknown);
        assert_eq!(build_state(Some(&json!({}))), BuildState::Unknown);
    }

    #[test]
    fn a_zero_crate_count_is_the_load_bearing_one() {
        let profile = json!({ "auto": { "crates": { "n": 0 } } });
        assert_eq!(build_state(Some(&profile)), BuildState::DocsOnly);
    }
}
